application-confluence-migrator-pro-default/pom.xml

```diff
@@ -29,6 +29,9 @@
   </parent>
   <artifactId>application-confluence-migrator-pro-default</artifactId>
   <name>Confluence Migrator Pro - Default API</name>
+  <properties>
+    <xwiki.extension.namespaces>{root}</xwiki.extension.namespaces>
```

Comment:
would there be any implication if the module is installed on root?

Reply:
We need to understand why the listener is not called. To be clear the use case is:

* ``application-confluence-migrator-pro-default`` is installed on the main wiki (at wiki level)
* this means an instance of ``ConfluenceFilteringListener`` (singleton) is registered in the Component Manager associated to the main wiki
* the user triggers a Confluence migration from a sub-wiki, which triggers a ``ConfluenceFilteringEvent``
* the Observation Manager (a core component, registered in the root component manager) needs to notify the listeners of this ``ConfluenceFilteringEvent``. So it needs to look for ``EventListener`` components that registered for this event. ATM the Observation Manager registers all event listeners (from all namespaces) in the same place, and they receive evens from all namespaces. So your listener registered on the main wiki should be notified of the event triggered in the subwiki.

Do you have any warning about event listener name collision in the logs? (i.e. if the confluence migration is install both in the main wiki and in the subwiki)

Reply:
After more testing, It seems that the listener works in the subwiki, but not for the main wiki. When I raised the issue, it worked the other way around.

After restarting the xwiki instance, I get these logs:
```
Listening for transport dt_socket at address: 5005
2023-10-17 10:48:16.689:INFO :oxtjl.NotifyListener:main: Starting Jetty on port 8080, on Java 11.0.20.1+1-post-Ubuntu-0ubuntu120.04 (OpenJDK Runtime Environment)...
2023-10-17 10:48:18,940 [main] INFO  iPropertiesConfigurationSource - Loading [xwiki.properties] from [file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/webapps/xwiki/WEB-INF/xwiki.properties] 
2023-10-17 10:48:19,003 [main] INFO  o.x.e.i.ServletEnvironment     - Using permanent directory [data] 
2023-10-17 10:48:20,256 [main] WARN  .x.e.m.i.MavenExtensionScanner - Collision between core extension [org.eclipse.jetty:jetty-client/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-client-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-client/pom.xml)] and [org.eclipse.jetty:jetty-client/9.4.49.v20220914 (file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/webapps/xwiki/WEB-INF/lib/jetty-client-9.4.49.v20220914.xed)] 
2023-10-17 10:48:20,256 [main] WARN  .x.e.m.i.MavenExtensionScanner - [org.eclipse.jetty:jetty-client/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-client-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-client/pom.xml)] is selected 
2023-10-17 10:48:20,258 [main] WARN  .x.e.m.i.MavenExtensionScanner - Collision between core extension [org.eclipse.jetty:jetty-util/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-util-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-util/pom.xml)] and [org.eclipse.jetty:jetty-util/9.4.49.v20220914 (file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/webapps/xwiki/WEB-INF/lib/jetty-util-9.4.49.v20220914.xed)] 
2023-10-17 10:48:20,258 [main] WARN  .x.e.m.i.MavenExtensionScanner - [org.eclipse.jetty:jetty-util/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-util-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-util/pom.xml)] is selected 
2023-10-17 10:48:20,264 [main] WARN  .x.e.m.i.MavenExtensionScanner - Collision between core extension [org.eclipse.jetty:jetty-http/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-http-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-http/pom.xml)] and [org.eclipse.jetty:jetty-http/9.4.49.v20220914 (file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/webapps/xwiki/WEB-INF/lib/jetty-http-9.4.49.v20220914.xed)] 
2023-10-17 10:48:20,264 [main] WARN  .x.e.m.i.MavenExtensionScanner - [org.eclipse.jetty:jetty-http/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-http-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-http/pom.xml)] is selected 
2023-10-17 10:48:20,266 [main] WARN  .x.e.m.i.MavenExtensionScanner - Collision between core extension [org.eclipse.jetty:jetty-io/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-io-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-io/pom.xml)] and [org.eclipse.jetty:jetty-io/9.4.49.v20220914 (file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/webapps/xwiki/WEB-INF/lib/jetty-io-9.4.49.v20220914.xed)] 
2023-10-17 10:48:20,266 [main] WARN  .x.e.m.i.MavenExtensionScanner - [org.eclipse.jetty:jetty-io/10.0.7 (jar:file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/jetty/lib/jetty-io-10.0.7.jar!/META-INF/maven/org.eclipse.jetty/jetty-io/pom.xml)] is selected 
2023-10-17 10:48:22,464 [main] INFO  o.x.s.s.i.EmbeddedSolr         - Updating Solr home directory at [/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/data/store/solr] 
2023-10-17 10:48:22,469 [main] INFO  o.x.s.s.i.EmbeddedSolr         - Starting embedded Solr server... 
2023-10-17 10:48:22,469 [main] INFO  o.x.s.s.i.EmbeddedSolr         - Using Solr home directory: [/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/data/store/solr] 
2023-10-17 10:48:22,851 [main] WARN  o.e.j.u.s.S.config             - Trusting all certificates configured for Client@2f229f60[provider=null,keyStore=null,trustStore=null] 
2023-10-17 10:48:22,852 [main] WARN  o.e.j.u.s.S.config             - No Client EndPointIdentificationAlgorithm configured for Client@2f229f60[provider=null,keyStore=null,trustStore=null] 
2023-10-17 10:48:22,937 [main] WARN  o.e.j.u.s.S.config             - Trusting all certificates configured for Client@1a48e48e[provider=null,keyStore=null,trustStore=null] 
2023-10-17 10:48:22,938 [main] WARN  o.e.j.u.s.S.config             - No Client EndPointIdentificationAlgorithm configured for Client@1a48e48e[provider=null,keyStore=null,trustStore=null] 
2023-10-17 10:48:22,976 [main] WARN  o.a.s.c.CoreContainer          - Not all security plugins configured!  authentication=disabled authorization=disabled.  Solr is only as secure as you make it. Consider configuring authentication/authorization before exposing Solr to users internal or external.  See https://s.apache.org/solrsecurity for more info 
2023-10-17 10:48:24,171 [main] INFO  o.x.s.s.i.EmbeddedSolr         - Started embedded Solr server. 
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.thoughtworks.xstream.converters.reflection.FieldDictionary (file:/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/webapps/xwiki/WEB-INF/lib/xstream-1.4.17.jar) to field java.util.Collections$UnmodifiableMap.serialVersionUID
WARNING: Please consider reporting this to the maintainers of com.thoughtworks.xstream.converters.reflection.FieldDictionary
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
2023-10-17 10:48:24,901 [main] INFO  o.x.s.f.i.FilesystemStoreTools - Using filesystem store directory [/home/teo/Work/xwiki-versions/xwiki-platform-distribution-flavor-jetty-hsqldb-14.10/data/store/file] 
2023-10-17 10:48:24,947 [main] WARN  .o.i.DefaultObservationManager - The [com.xwiki.confluencepro.internal.ConfluenceFilteringListener] listener is overwriting a previously registered listener [com.xwiki.confluencepro.internal.ConfluenceFilteringListener] since they both are registered under the same id [com.xwiki.confluencepro.internal.ConfluenceFilteringListener]. In the future consider removing a Listener first if you really want to register it again. 
2023-10-17 10:48:25.037:INFO :oxtjl.NotifyListener:main: ----------------------------------
2023-10-17 10:48:25.037:INFO :oxtjl.NotifyListener:main: Server started, you can now open http://teo-ROG-Zephyrus-M16-GU603ZEB-GU603ZE:8080/ in your browser to access your wiki.
2023-10-17 10:48:25.038:INFO :oxtjl.NotifyListener:main: ----------------------------------
2023-10-17 10:49:43,492 [XWiki initialization] INFO  .HibernateDataMigrationManager - Storage schema updates and data migrations are enabled 
2023-10-17 10:49:43,610 [XWiki initialization] INFO  .HibernateDataMigrationManager - No data migration to apply for wiki [xwiki] currently in version [140900000] 
2023-10-17 10:49:43,610 [XWiki initialization] INFO  .HibernateDataMigrationManager - Checking Hibernate mapping and updating schema if needed for wiki [xwiki] 
2023-10-17 10:49:43,790 [XWiki initialization] INFO  .HibernateDataMigrationManager - No data migration to apply for wiki [test] currently in version [140900000] 
2023-10-17 10:49:43,790 [XWiki initialization] INFO  .HibernateDataMigrationManager - Checking Hibernate mapping and updating schema if needed for wiki [test] 
2023-10-17 10:49:48,022 [solr/indexer job group daemon thread - org.xwiki.search.solr.internal.job.IndexerJob@6a0da002([solr, indexer])] INFO  o.x.s.s.i.j.IndexerJob         - Starting job of type [solr.indexer] with identifier [[solr, indexer]] 
2023-10-17 10:49:48,618 [solr/indexer job group daemon thread - org.xwiki.search.solr.internal.job.IndexerJob@6a0da002([solr, indexer])] INFO  o.x.s.s.i.j.IndexerJob         - 1 documents added, 0 deleted and 0 updated during the synchronization of the Solr index. 
2023-10-17 10:49:48,619 [solr/indexer job group daemon thread - org.xwiki.search.solr.internal.job.IndexerJob@6a0da002([solr, indexer])] INFO  o.x.s.s.i.j.IndexerJob         - Finished job of type [solr.indexer] with identifier [[solr, indexer]]
```
This line is of interest:

> 2023-10-17 10:48:24,947 [main] WARN  .o.i.DefaultObservationManager - The [com.xwiki.confluencepro.internal.ConfluenceFilteringListener] listener is overwriting a previously registered listener [com.xwiki.confluencepro.internal.ConfluenceFilteringListener] since they both are registered under the same id [com.xwiki.confluencepro.internal.ConfluenceFilteringListener]. In the future consider removing a Listener first if you really want to register it again.

Reply:
After some more debugging, it seems that when the `ConfluenceInputFilterStream` tries to notify the concerned listeners using the `ObservationManager`, it doesn't find the listener registered on the main wiki.
![image](https://github.com/xwikisas/application-confluence-migrator-pro/assets/45433221/24e9b57f-41a6-43fc-905f-2d7de6a74e7a)
This printscreen is taken inside the `ObservationManager.notify(Event.class)` function. It seems that the class of the passed event is different than the one registered in the observation manager. The classname is the same, but they are two different objects. The classloader also differs.

Reply:
Hmm, so the Confluence filter stream is also installed on the sub-wiki right? Then I think I understand the problem:

* you have a **single** event listener (whichever is registered last, between the main wiki one and the subwiki one)
* you have two ``ConfluenceFilteringEvent`` classes, one from the main wiki and one from the subwiki
* the Observation Manager binds the event listeners to the event class they listen to, which means the registered event listener is going to be bound either to the event from the main wiki or to the (same) event from the subwiki
* events of the class that wasn't bound are not "triggered" (well, they are but there's no listener bound to them).